File: ios/AccountCenter/Features/Login/LoginViewModel.swift
import SwiftUI

@MainActor
class LoginViewModel: ObservableObject {
    @Published var phoneNumber = ""
    @Published var password = ""
    @Published var verificationCode = ""
    @Published var loginMode: LoginMode = .password
    @Published var isLoading = false
    @Published var countdownSeconds = 0
    @Published var errorMessage: String?
    
    enum LoginMode {
        case password
        case verificationCode
    }
    
    private var countdownTimer: Timer?
    private let authManager: AuthManager
    
    init(authManager: AuthManager = .shared) {
        self.authManager = authManager
    }
    
    func login() async {
        guard !isLoading else { return }
        isLoading = true
        errorMessage = nil
        
        do {
            let request: LoginRequest
            switch loginMode {
            case .password:
                request = LoginRequest(phoneNumber: phoneNumber, password: password)
            case .verificationCode:
                request = LoginRequest(phoneNumber: phoneNumber, code: verificationCode)
            }
            
            let token = try await APIClient.shared.login(request: request)
            authManager.login(token: token)
        } catch let error as APIError {
            errorMessage = error.errorDescription
        } catch {
            errorMessage = "登录失败，请稍后重试"
        }
        
        isLoading = false
    }
    
    func sendVerificationCode() async {
        guard !isLoading, countdownSeconds == 0 else { return }
        isLoading = true
        errorMessage = nil
        
        do {
            let request = SMSSendRequest(phoneNumber: phoneNumber)
            try await APIClient.shared.sendSMS(request: request)
            startCountdown()
        } catch let error as APIError {
            errorMessage = error.errorDescription
        } catch {
            errorMessage = "发送失败，请稍后重试"
        }
        
        isLoading = false
    }
    
    private func startCountdown() {
        countdownSeconds = 60
        countdownTimer?.invalidate()
        
        countdownTimer = Timer.scheduledTimer(withTimeInterval: 1, repeats: true) { [weak self] _ in
            guard let self = self else { return }
            DispatchQueue.main.async {
                if self.countdownSeconds > 0 {
                    self.countdownSeconds -= 1
                } else {
                    self.countdownTimer?.invalidate()
                }
            }
        }
    }
}

File: android/app/src/main/java/com/accountcenter/ui/login/LoginViewModel.kt
package com.accountcenter.ui.login

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.accountcenter.model.LoginRequest
import com.accountcenter.model.SMSSendRequest
import com.accountcenter.repository.AuthRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class LoginUiState(
    val phoneNumber: String = "",
    val password: String = "",
    val verificationCode: String = "",
    val loginMode: LoginMode = LoginMode.PASSWORD,
    val isLoading: Boolean = false,
    val countdownSeconds: Int = 0,
    val errorMessage: String? = null
)

enum class LoginMode {
    PASSWORD, VERIFICATION_CODE
}

@HiltViewModel
class LoginViewModel @Inject constructor(
    private val authRepository: AuthRepository
) : ViewModel() {
    private val _uiState = MutableStateFlow(LoginUiState())
    val uiState: StateFlow<LoginUiState> = _uiState.asStateFlow()

    private var countdownJob: kotlinx.coroutines.Job? = null

    fun onPhoneNumberChange(value: String) {
        _uiState.value = _uiState.value.copy(phoneNumber = value, errorMessage = null)
    }

    fun onPasswordChange(value: String) {
        _uiState.value = _uiState.value.copy(password = value, errorMessage = null)
    }

    fun onVerificationCodeChange(value: String) {
        _uiState.value = _uiState.value.copy(verificationCode = value, errorMessage = null)
    }

    fun onLoginModeChange(mode: LoginMode) {
        _uiState.value = _uiState.value.copy(loginMode = mode)
    }

    fun login(onSuccess: () -> Unit) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            val state = _uiState.value
            val request = if (state.loginMode == LoginMode.PASSWORD) {
                LoginRequest.withPassword(state.phoneNumber, state.password)
            } else {
                LoginRequest.withCode(state.phoneNumber, state.verificationCode)
            }
            val result = authRepository.login(request)

            if (result.isSuccess) {
                onSuccess()
            } else {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    errorMessage = result.exceptionOrNull()?.message ?: "登录失败"
                )
            }
        }
    }

    fun sendVerificationCode() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            val request = SMSSendRequest(_uiState.value.phoneNumber)
            val result = authRepository.sendSMS(request)

            if (result.isSuccess) {
                startCountdown()
            } else {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    errorMessage = result.exceptionOrNull()?.message ?: "发送失败"
                )
            }
        }
    }

    private fun startCountdown() {
        _uiState.value = _uiState.value.copy(isLoading = false, countdownSeconds = 60)
        countdownJob?.cancel()
        countdownJob = viewModelScope.launch {
            while (_uiState.value.countdownSeconds > 0) {
                delay(1000)
                _uiState.value = _uiState.value.copy(countdownSeconds = _uiState.value.countdownSeconds - 1)
            }
        }
    }
}
